### PR TITLE
executors: Add site-admin dashboard

### LINKED ISSUE #27446: executors: Add site-admin dashboard
Add bare-bones site-admin dashboard just to show if executors have been connected to an instance recently. Should be used to check whether or not executor deployment has succeeded without setting up complete observability. Will eventually be fleshed out with additional context (running jobs, etc). Related to https://github.com/sourcegraph/sourcegraph/issues/26146.

### PR DESCRIPTION
**Pull request by efritz**
Fixes #27446. This PR adds a minimal dashboard to the site-admin UI that includes executor names, their (last reported) job counts, and the last time they were updated.

@eseliger has recently built out the other side of this effort in #27500. That PR sends additional telemetry data from the executor to the Sourcegraph instance so that we can track things like OS, versions, etc.

This PR does several things:

- Creates a new executor_heartbeats table to store the _last known_ heartbeat for unique executor instances
- Adds a store in internal/database to access/modify this data
- Writes to this store when heartbeat API calls are made in the executorqueue handler
- Exposes a GraphQL API to surface this data
- Adds a site-admin page with a filtered component over this data
- A periodic routine that deletes very old inactive executor instance records

A bit of executorqueue needed to be refactored to get the queue name from the route. This affected a few files of setup in the worker process.

### CODE REVIEW

File: client/web/src/enterprise/executors/CODENOTIFY [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# See https://github.com/sourcegraph/codenotify for documentation.
+
+**/* @efritz
```
Comment (eseliger) [inline client/web/src/enterprise/executors/CODENOTIFY:3]:
```suggestion
**/* @efritz
**/* @eseliger
```

File: internal/database/executors.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+	"github.com/keegancsmith/sqlf"
+
+	"github.com/sourcegraph/sourcegraph/internal/database/basestore"
+	"github.com/sourcegraph/sourcegraph/internal/database/dbutil"
+	"github.com/sourcegraph/sourcegraph/internal/timeutil"
+)
+
+// Executor describes an executor instance that has recently connected to Sourcegraph.
+type Executor struct {
```
Comment (eseliger) [inline internal/database/executors.go:16]: I think this should live in the types package

File: internal/database/executors.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	Offset int
+	Limit  int
+}
+
+// TODO - test
+// List returns a set of executor activity records matching the given options.
+//
+// 🚨 SECURITY: The caller must ensure that the actor is permitted to view executor details
+// (e.g., a site-admin).
+func (s *executorStore) List(ctx context.Context, opts ExecutorStoreListOptions) (_ []Executor, _ int, err error) {
```
Comment (eseliger) [inline internal/database/executors.go:106]: Do we anywhere else use this pattern of counting and listing in the same tx?

File: client/web/src/enterprise/site-admin/sidebaritems.ts [hunk 4]
```diff
@@ -98,11 +99,27 @@ const codeIntelGroup: SiteAdminSideBarGroup = {
     ],
 }
 
+const executorsGroup: SiteAdminSideBarGroup = {
```
Comment (eseliger) [inline client/web/src/enterprise/site-admin/sidebaritems.ts:102]: should this just be in maintenance?

File: client/web/src/enterprise/site-admin/sidebaritems.ts [hunk 5]
```diff
@@ -98,11 +99,27 @@ const codeIntelGroup: SiteAdminSideBarGroup = {
...
+        {
+            to: '/site-admin/executors',
+            label: 'Executors',
+        },
+    ],
+    // TODO - expand this to executors enabled when SSBC need this page
+    // as well. Right now we don't have an easy way to check if the
+    // executor accessToken is set in site-config, but that should be
+    // the condition of showing this.
+    condition: () => Boolean(window.context?.codeIntelAutoIndexingEnabled),
```
Comment (eseliger) [inline client/web/src/enterprise/site-admin/sidebaritems.ts:114]: I think we can always show it and have Rob design a nice small landing page that redirects to docs "WHY YOU SHOULD DO THIS"

**Review by eseliger — COMMENTED** [review]
This is all great!

I think the only missing critical piece to use this would be something to clean up executors that are older than X.

File: internal/database/executors.go (hunk 3 above)
Comment (efritz) [reply internal/database/executors.go:106]: All over codeintel.

File: client/web/src/enterprise/site-admin/sidebaritems.ts (hunk 5 above)
Comment (efritz) [reply client/web/src/enterprise/site-admin/sidebaritems.ts:114]: Yes, but maybe after it's less experimental? Rob will need our docs first 😛

File: client/web/src/enterprise/site-admin/sidebaritems.ts (hunk 4 above)
Comment (efritz) [reply client/web/src/enterprise/site-admin/sidebaritems.ts:102]: Could be, but I'd imagine we'd add more things to this in the future. Would be easy to move.

File: client/web/src/enterprise/site-admin/sidebaritems.ts (hunk 5 above)
Comment (eseliger) [reply client/web/src/enterprise/site-admin/sidebaritems.ts:114]: sure, I think I was just gonna say maybe it's alright to move it there now 🤷‍♂️

File: client/web/src/enterprise/site-admin/sidebaritems.ts (hunk 4 above)
Comment (efritz) [reply client/web/src/enterprise/site-admin/sidebaritems.ts:102]: Fixing.

File: internal/database/executors.go (hunk 2 above)
Comment (efritz) [reply internal/database/executors.go:16]: Fixing.

**Comment by efritz** [comment]
@eseliger 

> I think the only missing critical piece to use this would be something to clean up executors that are older than X.

Plan to add pre-non-draft.

File: enterprise/cmd/frontend/internal/executorqueue/queues/batches/queue.go [hunk 6]
```diff
@@ -23,6 +23,7 @@ func QueueOptions(db dbutil.DB, accessToken func() string, observationContext *o
 
 	store := background.NewBatchSpecWorkspaceExecutionWorkerStore(basestore.NewHandleWithDB(db, sql.TxOptions{}), observationContext)
 	return handler.QueueOptions{
+		Name:                   "codeintel",
```
Comment (efritz) [inline enterprise/cmd/frontend/internal/executorqueue/queues/batches/queue.go:26]:
```suggestion
		Name:                   "batches",
```

File: enterprise/cmd/worker/internal/executors/janitor_config.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+	env.BaseConfig
+
+	CleanupTaskInterval    time.Duration
+	HeartbeatRecordsMaxAge time.Duration
+}
+
+var janitorConfigInst = &janitorConfig{}
+
+func (c *janitorConfig) Load() {
+	c.CleanupTaskInterval = c.GetInterval("EXECUTOR_CLEANUP_TASK_INTERVAL", "30m", "The frequency with which to run executor cleanup tasks.")
```
Comment (efritz) [inline enterprise/cmd/worker/internal/executors/janitor_config.go:19]:
```suggestion
	c.CleanupTaskInterval = c.GetInterval("EXECUTORS_CLEANUP_TASK_INTERVAL", "30m", "The frequency with which to run executor cleanup tasks.")
`

File: enterprise/cmd/worker/internal/executors/janitor_config.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+	CleanupTaskInterval    time.Duration
+	HeartbeatRecordsMaxAge time.Duration
+}
+
+var janitorConfigInst = &janitorConfig{}
+
+func (c *janitorConfig) Load() {
+	c.CleanupTaskInterval = c.GetInterval("EXECUTORS_CLEANUP_TASK_INTERVAL", "30m", "The frequency with which to run executor cleanup tasks.")
+
```
Comment (efritz) [inline enterprise/cmd/worker/internal/executors/janitor_config.go:20]:
```suggestion
```

**Review by eseliger — APPROVED** [review]

**Pull request merged**